File: 437/path-sum-iii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    HashMap<Long,Integer>map = new HashMap<>();
    Stack<Long> stack = new Stack<>(); //prefix sum
    int ans = 0;

    public int pathSum(TreeNode root, int targetSum) {
        //buildPrefixSum(root, 0);
        travel(root, targetSum, 0);
        return ans;
    }
    void travel(TreeNode root, long targetSum, long sum) {
        if(root==null) return;
        long now = sum + root.val;
        stack.push(now);

        if(now==targetSum) ans++;
        long diff = now-targetSum;
        if(map.containsKey(diff)){
            ans += map.get(diff);
        }
        if(map.containsKey(now)){
            map.put(now, map.get(now)+1);
        }else map.put(now, 1); //加

        travel(root.left, targetSum, now);
        travel(root.right, targetSum, now);
        map.put(now, map.get(now)-1);//減

        stack.pop();
    }
/*    void buildPrefixSum(TreeNode root, int sum) {
        if(root==null) return;
        root.val += sum; //糟! int 不夠裝 20億，要改用long 吧! 要換資料結栟
        if(root.left!=null) buildPrefixSum(root.left, root.val);
        if(root.right!=null) buildPrefixSum(root.right, root.val);
    }*/
}//case 127/128: [1000000000,1000000000,null,294967296,null,1000000000,null,1000000000,null,1000000000] 0

File: 437/path-sum-iii.py
# LeetCode 437. Path Sum III
# 想知道有幾條 path sum 加起來剛好是 targetSum
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        preSum = 0  # 利用 preSum 記錄 root 到 node 的 sum 值
        counter = Counter()  # 利用 counter 記錄「之前各種 preSum 出現的次數
        counter[0] += 1  # 最基礎的 case，讓 root 開始的 preSum 也有機會==targetSum
        def helper(root, preSum):  # 利用「函式呼叫函式」把 tree 全部走過一次
            if root==None: return 0  # 末端結束
            preSum += root.val  # 加入現在node的值
            ans = counter[preSum-targetSum]  # 到此 node 結束的 path 有幾種可能？
            
            counter[preSum] += 1  # 兩個小孩，都在「現在的基礎」出發
            ans += helper(root.left, preSum) + helper(root.right, preSum)
            counter[preSum] -= 1  # 算完後，再扣掉「現在的基礎」
            return ans
        return helper(root, 0)
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
# case 120/130: root = [1], targetSum = 0
